File: test-harness/tests/resources/r2_contract_server.py
#!/usr/bin/env python3
"""Loopback S3 contract fixture. Validates SigV4 and rejects R2-unsupported ACLs.

This is a protocol test, not proof of a real R2 account's permissions or TLS setup.
Only synthetic credentials and in-memory objects are used.
"""
import hashlib
import hmac
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
import json
import os
import threading
import time
from urllib.parse import parse_qsl, quote, unquote, urlsplit

ACCESS = 'contract-access'
SECRET = 'contract-secret'
OBJECTS = {}
EVENTS = []


def digest(value):
    return hashlib.sha256(value).hexdigest()


def signature(method, target, headers, body):
    parsed = urlsplit(target)
    query = dict(parse_qsl(parsed.query, keep_blank_values=True))
    presigned = 'X-Amz-Signature' in query
    if presigned:
        credential = query['X-Amz-Credential']
        signed = query['X-Amz-SignedHeaders']
        supplied = query.pop('X-Amz-Signature')
        stamp = query['X-Amz-Date']
        payload = 'UNSIGNED-PAYLOAD'
        assert 1 <= int(query['X-Amz-Expires']) <= 900, 'unbounded signed URL'
    else:
        algorithm, auth = headers['authorization'].split(' ', 1)
        assert algorithm == 'AWS4-HMAC-SHA256'
        fields = dict(part.strip().split('=', 1) for part in auth.split(','))
        credential, signed, supplied = fields['Credential'], fields['SignedHeaders'], fields['Signature']
        stamp = headers['x-amz-date']
        payload = headers.get('x-amz-content-sha256', digest(body))
        assert payload == digest(body), 'body hash mismatch'
    access, date, region, service, terminator = credential.split('/')
    assert (access, service, terminator) == (ACCESS, 's3', 'aws4_request')
    assert region in ('auto', '', 'us-east-1'), 'unsupported R2 signing region'
    canonical_uri = quote(unquote(parsed.path), safe='/~')
    canonical_query = '&'.join(f'{quote(k, safe="~")}={quote(v, safe="~")}' for k, v in sorted(query.items()))
    canonical_headers = ''.join(f'{name}:{" ".join(headers[name].strip().split())}\n' for name in signed.split(';'))
    canonical = '\n'.join([method, canonical_uri, canonical_query, canonical_headers, signed, payload])
    scope = '/'.join([date, region, service, terminator])
    text = '\n'.join(['AWS4-HMAC-SHA256', stamp, scope, digest(canonical.encode())])
    signing = ('AWS4' + SECRET).encode()
    for part in (date, region, service, terminator):
        signing = hmac.new(signing, part.encode(), hashlib.sha256).digest()
    expected = hmac.new(signing, text.encode(), hashlib.sha256).hexdigest()
    assert hmac.compare_digest(expected, supplied), 'signature mismatch'
    return unquote(parsed.path)


class Handler(BaseHTTPRequestHandler):
    protocol_version = 'HTTP/1.1'

    def log_message(self, *_):
        pass

    def respond(self, status, body=b'', headers=None):
        self.send_response(status)
        for name, value in (headers or {}).items():
            self.send_header(name, value)
        self.send_header('Content-Length', str(len(body)))
        self.end_headers()
        if self.command != 'HEAD':
            self.wfile.write(body)

    def handle_request(self):
        if self.path == '/__evidence':
            return self.respond(200, json.dumps(EVENTS).encode(), {'Content-Type': 'application/json'})
        body = self.rfile.read(int(self.headers.get('Content-Length', '0')))
        headers = {key.lower(): value for key, value in self.headers.items()}
        event = {'method': self.command, 'path': self.path.split('?')[0], 'valid': False}
        EVENTS.append(event)
        try:
            assert 'x-amz-acl' not in headers and 'acl' not in dict(parse_qsl(urlsplit(self.path).query, keep_blank_values=True)), 'ACL unsupported'
            key = signature(self.command, self.path, headers, body)
            assert key.startswith(('/private-contract/', '/public-contract/')), 'wrong bucket or addressing style'
            event['valid'] = True
        except Exception as error:
            event['error'] = str(error)
            return self.respond(403, ('<Error><Code>AccessDenied</Code><Message>' + str(error) + '</Message></Error>').encode(), {'Content-Type': 'application/xml'})
        if self.command == 'PUT' and 'x-amz-copy-source' in headers:
            if key.endswith('/fail-copy.pdf'):
                return self.respond(403, b'<Error><Code>AccessDenied</Code><Message>copy denied</Message></Error>', {'Content-Type': 'application/xml'})
            source = unquote(headers['x-amz-copy-source'])
            if source not in OBJECTS:
                return self.respond(404)
            OBJECTS[key] = OBJECTS[source]
            return self.respond(200, b'<CopyObjectResult><ETag>"copied"</ETag></CopyObjectResult>', {'Content-Type': 'application/xml'})
        if self.command == 'PUT':
            OBJECTS[key] = (body, headers.get('content-type', 'application/octet-stream'))
            return self.respond(200, headers={'ETag': '"' + hashlib.md5(body).hexdigest() + '"'})
        if self.command == 'DELETE':
            OBJECTS.pop(key, None)
            return self.respond(204)
        if key not in OBJECTS:
            return self.respond(404)
        body, content_type = OBJECTS[key]
        self.respond(200, body, {'Content-Type': content_type, 'ETag': '"' + hashlib.md5(body).hexdigest() + '"',
                                 'Last-Modified': 'Thu, 10 Sep 2026 00:00:00 GMT'})

    do_HEAD = handle_request
    do_GET = handle_request
    do_PUT = handle_request
    do_DELETE = handle_request


def main():
    parent = os.getppid()
    server = ThreadingHTTPServer(('127.0.0.1', 0), Handler)
    def watch_parent():
        while os.getppid() == parent:
            time.sleep(1)
        server.shutdown()
    threading.Thread(target=watch_parent, daemon=True).start()
    print(server.server_port, flush=True)
    server.serve_forever(poll_interval=0.2)
    server.server_close()


if __name__ == '__main__':
    main()

File: test-harness/tests/resources/R2ContractServer.java
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;
import java.net.InetSocketAddress;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Executors;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

/** Synthetic, loopback-only S3 protocol fixture; never accepts real credentials. */
class R2ContractServer {
    private static final String ACCESS = "contract-access";
    private static final String SECRET = "contract-secret";
    private static final Map<String, StoredObject> OBJECTS = new ConcurrentHashMap<>();
    private static final List<Event> EVENTS = java.util.Collections.synchronizedList(new ArrayList<>());

    record StoredObject(byte[] body, String contentType) {}
    record Event(String method, String path, boolean valid, String error) {}

    public static void main(String[] args) throws Exception {
        var server = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);
        server.createContext("/", R2ContractServer::handle);
        server.setExecutor(Executors.newCachedThreadPool());
        server.start();
        System.out.println(server.getAddress().getPort());
        System.out.flush();
        // The CommandBox task or TestBox spec owns this child process.
        Thread.currentThread().join();
    }

    private static void handle(HttpExchange exchange) throws java.io.IOException {
        try {
            if (exchange.getRequestURI().getPath().equals("/__evidence")) {
                respond(exchange, 200, evidence().getBytes(StandardCharsets.UTF_8), "application/json", null);
                return;
            }
            byte[] body = exchange.getRequestBody().readAllBytes();
            var headers = new HashMap<String, String>();
            exchange.getRequestHeaders().forEach((name, values) ->
                headers.put(name.toLowerCase(java.util.Locale.ROOT), values.get(0)));
            var query = parseQuery(exchange.getRequestURI().getRawQuery());
            String method = exchange.getRequestMethod();
            String path = exchange.getRequestURI().getPath();
            try {
                require(!headers.containsKey("x-amz-acl") && !query.containsKey("acl"), "ACL unsupported");
                verifySignature(exchange, headers, query, body);
                require(path.startsWith("/private-contract/") || path.startsWith("/public-contract/"),
                    "wrong bucket or addressing style");
                EVENTS.add(new Event(method, path, true, null));
            } catch (Exception error) {
                EVENTS.add(new Event(method, path, false, error.getMessage()));
                respond(exchange, 403,
                    ("<Error><Code>AccessDenied</Code><Message>" + error.getMessage()
                        + "</Message></Error>").getBytes(StandardCharsets.UTF_8),
                    "application/xml", null);
                return;
            }

            if (method.equals("PUT") && headers.containsKey("x-amz-copy-source")) {
                if (path.endsWith("/fail-copy.pdf")) {
                    respond(exchange, 403,
                        "<Error><Code>AccessDenied</Code><Message>copy denied</Message></Error>"
                            .getBytes(StandardCharsets.UTF_8),
                        "application/xml", null);
                    return;
                }
                var source = decode(headers.get("x-amz-copy-source"));
                var object = OBJECTS.get(source);
                if (object == null) {
                    respond(exchange, 404, new byte[0], null, null);
                    return;
                }
                OBJECTS.put(path, object);
                respond(exchange, 200,
                    "<CopyObjectResult><ETag>\"copied\"</ETag></CopyObjectResult>"
                        .getBytes(StandardCharsets.UTF_8),
                    "application/xml", null);
                return;
            }
            if (method.equals("PUT")) {
                OBJECTS.put(path, new StoredObject(body,
                    headers.getOrDefault("content-type", "application/octet-stream")));
                respond(exchange, 200, new byte[0], null, "\"" + md5(body) + "\"");
                return;
            }
            if (method.equals("DELETE")) {
                OBJECTS.remove(path);
                respond(exchange, 204, new byte[0], null, null);
                return;
            }
            var object = OBJECTS.get(path);
            if (object == null) {
                respond(exchange, 404, new byte[0], null, null);
                return;
            }
            respond(exchange, 200, object.body(), object.contentType(),
                "\"" + md5(object.body()) + "\"");
        } finally {
            exchange.close();
        }
    }

    private static void verifySignature(HttpExchange exchange, Map<String, String> headers,
                                        Map<String, String> query, byte[] body) throws Exception {
        String credential, signed, supplied, stamp, payload;
        if (query.containsKey("X-Amz-Signature")) {
            credential = query.get("X-Amz-Credential");
            signed = query.get("X-Amz-SignedHeaders");
            supplied = query.remove("X-Amz-Signature");
            stamp = query.get("X-Amz-Date");
            payload = "UNSIGNED-PAYLOAD";
            int seconds = Integer.parseInt(query.get("X-Amz-Expires"));
            require(seconds >= 1 && seconds <= 900, "unbounded signed URL");
        } else {
            var auth = headers.get("authorization");
            require(auth != null && auth.startsWith("AWS4-HMAC-SHA256 "), "wrong signing algorithm");
            var fields = new HashMap<String, String>();
            for (var part : auth.substring(17).split(",")) {
                var pair = part.trim().split("=", 2);
                fields.put(pair[0], pair[1]);
            }
            credential = fields.get("Credential");
            signed = fields.get("SignedHeaders");
            supplied = fields.get("Signature");
            stamp = headers.get("x-amz-date");
            payload = headers.getOrDefault("x-amz-content-sha256", sha256(body));
            require(payload.equals(sha256(body)), "body hash mismatch");
        }
        var scopeParts = credential.split("/", -1);
        require(scopeParts.length == 5, "invalid credential scope");
        var access = scopeParts[0];
        var date = scopeParts[1];
        var region = scopeParts[2];
        var service = scopeParts[3];
        var terminator = scopeParts[4];
        require(access.equals(ACCESS) && service.equals("s3")
            && terminator.equals("aws4_request"), "invalid credential scope");
        require(region.equals("auto") || region.isEmpty() || region.equals("us-east-1"),
            "unsupported R2 signing region");
        var canonicalQuery = new ArrayList<String>();
        query.forEach((key, value) -> canonicalQuery.add(awsEncode(key) + "=" + awsEncode(value)));
        canonicalQuery.sort(Comparator.naturalOrder());
        var canonicalHeaders = new StringBuilder();
        for (var name : signed.split(";")) {
            var value = headers.get(name);
            require(value != null, "missing signed header");
            canonicalHeaders.append(name).append(":")
                .append(String.join(" ", value.trim().split("\\s+"))).append("\n");
        }
        var canonical = String.join("\n", exchange.getRequestMethod(),
            awsEncodePath(decode(exchange.getRequestURI().getRawPath())),
            String.join("&", canonicalQuery), canonicalHeaders.toString(), signed, payload);
        var scope = String.join("/", date, region, service, terminator);
        var toSign = String.join("\n", "AWS4-HMAC-SHA256", stamp, scope,
            sha256(canonical.getBytes(StandardCharsets.UTF_8)));
        byte[] signing = ("AWS4" + SECRET).getBytes(StandardCharsets.UTF_8);
        for (var part : Arrays.asList(date, region, service, terminator)) {
            signing = hmac(signing, part);
        }
        require(MessageDigest.isEqual(hex(hmac(signing, toSign)).getBytes(StandardCharsets.US_ASCII),
            supplied.getBytes(StandardCharsets.US_ASCII)), "signature mismatch");
    }

    private static Map<String, String> parseQuery(String raw) {
        var result = new HashMap<String, String>();
        if (raw != null) {
            for (var item : raw.split("&")) {
                var pair = item.split("=", 2);
                result.put(decode(pair[0]), pair.length == 2 ? decode(pair[1]) : "");
            }
        }
        return result;
    }

    private static String decode(String encoded) {
        return URLDecoder.decode(encoded.replace("+", "%2B"), StandardCharsets.UTF_8);
    }

    private static String awsEncodePath(String value) {
        return awsEncode(value).replace("%2F", "/");
    }

    private static String awsEncode(String value) {
        var output = new StringBuilder();
        for (byte octet : value.getBytes(StandardCharsets.UTF_8)) {
            int c = octet & 0xff;
            if (c >= 'A' && c <= 'Z' || c >= 'a' && c <= 'z'
                || c >= '0' && c <= '9' || c == '-' || c == '_' || c == '.' || c == '~') {
                output.append((char) c);
            } else {
                output.append('%').append(String.format("%02X", c));
            }
        }
        return output.toString();
    }

    private static byte[] hmac(byte[] key, String value) throws Exception {
        var mac = Mac.getInstance("HmacSHA256");
        mac.init(new SecretKeySpec(key, "HmacSHA256"));
        return mac.doFinal(value.getBytes(StandardCharsets.UTF_8));
    }

    private static String sha256(byte[] bytes) throws Exception {
        return hex(MessageDigest.getInstance("SHA-256").digest(bytes));
    }

    private static String md5(byte[] bytes) {
        try {
            return hex(MessageDigest.getInstance("MD5").digest(bytes));
        } catch (Exception error) {
            throw new IllegalStateException(error);
        }
    }

    private static String hex(byte[] bytes) {
        return java.util.HexFormat.of().formatHex(bytes);
    }

    private static void require(boolean condition, String message) {
        if (!condition) throw new IllegalArgumentException(message);
    }

    private static String evidence() {
        var result = new StringBuilder("[");
        synchronized (EVENTS) {
            for (var event : EVENTS) {
                if (result.length() > 1) result.append(",");
                result.append("{\"method\":\"").append(event.method())
                    .append("\",\"path\":\"").append(event.path())
                    .append("\",\"valid\":").append(event.valid());
                if (event.error() != null) {
                    result.append(",\"error\":\"").append(event.error()).append("\"");
                }
                result.append("}");
            }
        }
        return result.append("]").toString();
    }

    private static void respond(HttpExchange exchange, int status, byte[] body,
                                String type, String etag) throws java.io.IOException {
        if (type != null) exchange.getResponseHeaders().set("Content-Type", type);
        if (etag != null) exchange.getResponseHeaders().set("ETag", etag);
        if (status == 200 && etag != null && body.length > 0) {
            exchange.getResponseHeaders().set("Last-Modified", "Thu, 10 Sep 2026 00:00:00 GMT");
        }
        if (status == 204) {
            exchange.sendResponseHeaders(status, -1);
            return;
        }
        exchange.getResponseHeaders().set("Content-Length", String.valueOf(body.length));
        exchange.sendResponseHeaders(status, exchange.getRequestMethod().equals("HEAD") ? -1 : body.length);
        if (!exchange.getRequestMethod().equals("HEAD")) {
            exchange.getResponseBody().write(body);
        }
    }
}
